qt/applications/workbench/workbench/app/mainwindow.py

```diff
@@ -414,10 +424,54 @@ def open_file(self):
             return
         self.editor.open_file_in_new_tab(filepath)
 
-    def save_file(self):
-        # todo: how should this interact with project saving and workspaces when they are implemented?
+    def save_script(self):
         self.editor.save_current_file()
 
+    def save_project(self):
+        if self.last_project_location is None:
+            self.save_project_as()
+        else:
+            # Clear directory before saving to remove old workspaces
+            files = glob.glob(self.last_project_location + '/.*')
+            for f in files:
+                os.remove(f)
+            # Actually save
+            workspaces_to_save = AnalysisDataService.getObjectNames()
+            project_saver = projectsaver.ProjectSaver()
+            project_saver.save_project(directory=self.last_project_location, workspace_to_save=workspaces_to_save,
+                                       interfaces_to_save=None)
+
+    def save_project_as(self):
+        directory = ""
+        # Check if it exists
+        first_pass = True
+        while first_pass or (not os.path.isdir(directory) or not os.path.exists(directory)):
```

Comment:
Could you help me understand the purpose of this loop? Doesn't the `QFileDialog` ensure that we get back an existing directory?

Reply:
It isn't actually doing what was intended. Basically it's a do while loop that checks the directory is either new or is definitely a project directory. (At least on the most recent commits)